File: frontend/src/services/terminal.ts
import { Events } from "@wailsio/runtime";
import { TerminalService } from "../../bindings/pi-desk/internal/appservice";
import type { TerminalState } from "../../bindings/pi-desk/internal/domain";

export interface TerminalEvent {
  threadId: string;
  /** Which terminal of the task produced this. Empty means the task's one and only terminal, the
   *  shape the runtime used before a task could own several. */
  sessionId?: string;
  type: "output" | "error" | "exit";
  generation?: number;
  sequence: number;
  dataB64?: string;
  exitCode?: number;
  error?: string;
  /** Wall-clock millisecond at which the pseudo-terminal produced `dataB64`. */
  emittedAt?: number;
}

export type TerminalWorkspaceReference = string | { workspaceId: string };

export const terminalService = {
  start(threadId: string, sessionId: string | undefined, workspace: TerminalWorkspaceReference, columns: number, rows: number): Promise<TerminalState> {
    const reference = typeof workspace === "string" ? { workspacePath: workspace } : workspace;
    return TerminalService.Start({ threadId, sessionId, ...reference, columns, rows });
  },
  snapshot(threadId: string, sessionId?: string, workspaceId?: string): Promise<TerminalState> {
    return TerminalService.Snapshot({ threadId, sessionId, workspaceId });
  },
  write(threadId: string, sessionId: string | undefined, data: string): Promise<void> {
    return TerminalService.Write({ threadId, sessionId, data });
  },
  resize(threadId: string, sessionId: string | undefined, columns: number, rows: number): Promise<void> {
    return TerminalService.Resize({ threadId, sessionId, columns, rows });
  },
  stop(threadId: string, sessionId?: string): Promise<void> {
    return TerminalService.Stop({ threadId, sessionId });
  },
};

export function onTerminalEvent(callback: (event: TerminalEvent) => void): () => void {
  return Events.On("terminal:event", (event) => callback(event.data as TerminalEvent));
}

File: internal/domain/terminal.go
package domain

type StartTerminalRequest struct {
	ThreadID      string `json:"threadId"`
	SessionID     string `json:"sessionId,omitempty"`
	WorkspaceID   string `json:"workspaceId,omitempty"`
	WorkspacePath string `json:"workspacePath,omitempty"`
	Columns       int    `json:"columns"`
	Rows          int    `json:"rows"`
}

type TerminalRequest struct {
	ThreadID    string `json:"threadId"`
	SessionID   string `json:"sessionId,omitempty"`
	WorkspaceID string `json:"workspaceId,omitempty"`
}

type TerminalWriteRequest struct {
	ThreadID  string `json:"threadId"`
	SessionID string `json:"sessionId,omitempty"`
	Data      string `json:"data"`
}

type TerminalResizeRequest struct {
	ThreadID  string `json:"threadId"`
	SessionID string `json:"sessionId,omitempty"`
	Columns   int    `json:"columns"`
	Rows      int    `json:"rows"`
}

type TerminalState struct {
	ThreadID   string `json:"threadId"`
	SessionID  string `json:"sessionId,omitempty"`
	CWD        string `json:"cwd,omitempty"`
	Shell      string `json:"shell,omitempty"`
	Running    bool   `json:"running"`
	Generation uint64 `json:"generation,omitempty"`
	Sequence   uint64 `json:"sequence"`
	OutputB64  string `json:"outputB64,omitempty"`
}

type TerminalEvent struct {
	ThreadID   string `json:"threadId"`
	SessionID  string `json:"sessionId,omitempty"`
	Type       string `json:"type"`
	Generation uint64 `json:"generation,omitempty"`
	Sequence   uint64 `json:"sequence"`
	DataB64    string `json:"dataB64,omitempty"`
	ExitCode   int    `json:"exitCode,omitempty"`
	Error      string `json:"error,omitempty"`
}
